iris-mpc-cpu/src/execution/hawk_main.rs

```diff
@@ -0,0 +1,174 @@
+    #[clap(short, long)]
+    party_index: usize,
+}
+
+pub async fn hawk_main(args: HawkArgs) -> Result<()> {
+    // ---- Shared setup ----
+
+    let n_parties = 3;
+    let port_start = 40000;
+    let search_params = HnswSearcher::default();
+
+    let identities = generate_local_identities();
+
+    let addresses = (0..n_parties)
+        .map(|i| format!("127.0.0.1:{}", port_start + i))
+        .collect_vec();
+
+    let role_assignments: RoleAssignment = identities
+        .iter()
+        .enumerate()
+        .map(|(index, id)| (Role::new(index), id.clone()))
+        .collect();
+
+    // ---- My network setup ----
+
+    let my_index = args.party_index;
+    let my_identity = identities[my_index].clone();
+    let my_address = addresses[my_index].clone();
+
+    println!("🦅 Starting Hawk node {my_index}");
+
+    let grpc_config = GrpcConfig {
+        timeout_duration: Duration::from_secs(1),
+    };
+
+    let player = GrpcNetworking::new(my_identity.clone(), grpc_config);
+
+    // Start server.
+    {
+        let player = player.clone();
+        let socket = my_address.parse().unwrap();
+        tokio::spawn(async move {
+            Server::builder()
+                .add_service(PartyNodeServer::new(player))
+                .serve(socket)
+                .await
+                .unwrap();
+        });
+    }
+
+    // TODO: Retry until all servers are up.
+    tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+
+    // Connect to other players.
+    izip!(&identities, &addresses)
+        .filter(|(_, address)| address != &&my_address)
+        .map(|(identity, address)| {
+            let player = player.clone();
+            let identity = identity.clone();
+            let url = format!("http://{}", address);
+            async move { player.connect_to_party(identity, &url).await }
+        })
+        .collect::<JoinSet<_>>()
+        .join_all()
+        .await
+        .into_iter()
+        .collect::<Result<()>>()?;
+
+    // ---- My state ----
+    // TODO: Persistence.
+
+    let iris_store = SharedIrisesRef::default();
+    let mut graph_store = GraphMem::<Aby3Store>::new();
+    let mut graph_rng = AesRng::seed_from_u64(123);
+
+    // ---- MPC session ----
+    // TODO: Manage parallel sessions.
+
+    let session_id = SessionId::from(0_u64);
+    let my_session_seed = [0_u8; 16];
+
+    player.create_session(session_id).await?;
+
+    let boot_session = BootSession {
+        session_id,
+        role_assignments: Arc::new(role_assignments.clone()),
+        networking: Arc::new(player.clone()),
+        own_identity: my_identity.clone(),
+    };
+
+    let prf = setup_replicated_prf(&boot_session, my_session_seed).await?;
+
+    let session = Session {
+        boot_session,
+        setup: prf,
+    };
+
+    let mut aby3_store = Aby3Store {
+        session,
+        storage: iris_store,
+        owner: my_identity,
+    };
+    assert_eq!(aby3_store.get_owner_index(), my_index);
```

Comment:
Shall it return `Err` perhaps?

Reply:
It’s more an assertion that I understood the method correctly.